### ISSUE TITLE
Fail to copy Roslyn compiler bits to the right location on build server

### ISSUE THREAD

**Issue by Jinhuafei** [aspnet/RoslynCodeDomProvider#2]
See [SO issue](http://stackoverflow.com/questions/32780315/could-not-find-a-part-of-the-path-bin-roslyn-csc-exe).

**Comment by vascofernandes** [comment]
Also affected by this problem.

**Comment by niemyjski** [comment]
I'm also affected by this. I need to be able to control the path.

**Comment by ejsmith** [comment]
Actually, my question is why do we even need to shell csc.exe in the 1st place? Why can't we just use the Roslyn API to compile the code?

**Comment by Jinhuafei** [comment]
@ejsmith DotNetCompilerPlatform is for runtime compilation(e.g. aspx/cshtml/global.asax etc..). It is implemented as a [language provider](https://msdn.microsoft.com/en-us/library/k6bttwes(v=vs.110).aspx) which is the only way asp.net provides to do the runtime compilation.

**Comment by niemyjski** [comment]
@Jinhuafei Yes, we get that, but it's implemented as a code Dom provider which just shells the compiler with command line arguments. What we are asking is why doesn't the code dom provider just compile in process using roslyn instead of shelling it.

**Comment by davidfowl** [comment]
The problem was controlling the memory allocated by the compiler in certain cases. There are some generated code patterns that the aspx code generator generates that cause the Roslyn compiler to either stackoverflow or run out of memory parsing. The only way to fix that was to run then compiler out of process (since there were no changes being made to code gen)

**Comment by ejsmith** [comment]
@davidfowl that makes sense. Is ASPNET Core still taking the out of process approach as well then?

**Comment by Jinhuafei** [comment]
David is right. We tried both, and out-proc compilation model works better. Besides what David mentioned, we also saw working set increment in w3wp process when using in-proc compilation.

**Comment by davidfowl** [comment]
> @davidfowl that makes sense. Is ASPNET Core still taking the out of process approach as well then?

We discussed it but haven't built in the infrastructure to make that work as well as ASP.NET (see ASP.NET Temporary Files). We also changed some of the code generation patterns in Razor to help the issue a bit. On top of that, we also now pre-compile views by default in on publish so your production site should never really be using roslyn at runtime.
